core/preconditioner/isai.cpp

```diff
@@ -0,0 +1,146 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/preconditioner/isai.hpp>
+
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/preconditioner/isai_kernels.hpp"
+
+
+namespace gko {
+namespace preconditioner {
+namespace isai {
+
+
+GKO_REGISTER_OPERATION(generate_l_inverse, isai::generate_l_inverse);
+GKO_REGISTER_OPERATION(generate_u_inverse, isai::generate_u_inverse);
+
+
+}  // namespace isai
+
+
+/**
+ * Helper function that converts the given matrix to the (const) CSR format with
+ * additional sorting.
+ *
+ * If the given matrix was already sorted, is on the same executor and with a
+ * dynamic type of `const Csr`, the same pointer is returned with an empty
+ * deleter.
+ * In all other cases, a new matrix is created, which stores the converted Csr
+ * matrix.
+ * If `skip_sorting` is false, the matrix will be sorted by column index,
+ * otherwise, it will not be sorted.
+ */
+template <typename Csr>
+std::unique_ptr<const Csr, std::function<void(const Csr *)>>
```

Comment:
This can probably marked as `@internal` or in case it is useful for other cases, then moved to a common file ?

Reply:
I did not think about marking this as `@internal` because it is in the `.cpp` file, but I will add it.